request-chroot.go

```diff
@@ -0,0 +1,341 @@
+// +build !plan9
+
+package sftp
+
+// This serves as an example of how to implement the request server handler as
+// well as a dummy backend for testing. It implements an in-memory backend that
+// works as a very simple filesystem with simple flat key-value lookup system.
+
+import (
+	"errors"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+)
+
+// ChrootHandler returns a Hanlders object with the specific root path.
+func ChrootHandler(rootPath string) Handlers {
+	chroot := &chroot{
+		rootPath: rootPath,
+	}
+	return Handlers{chroot, chroot, chroot, chroot}
+}
+
+type chroot struct {
+	rootPath string
+}
+
+func (fs *chroot) toOsFlags(flags uint32) (int, error) {
+	var osFlags int
+	if flags&sshFxfRead != 0 || flags&sshFxfWrite != 0 {
+		osFlags |= os.O_RDWR
+	} else if flags&sshFxfWrite != 0 {
+		osFlags |= os.O_WRONLY
+	} else if flags&sshFxfRead != 0 {
+		osFlags |= os.O_RDONLY
+	} else {
+		// how are they opening?
+		return 0, syscall.EINVAL
+	}
+	if flags&sshFxfCreat != 0 {
+		osFlags |= os.O_CREATE
+	}
+	if flags&sshFxfTrunc != 0 {
+		osFlags |= os.O_TRUNC
+	}
+	if flags&sshFxfExcl != 0 {
+		osFlags |= os.O_EXCL
+	}
+	return osFlags, nil
+}
+
+const OUT_PATH_PREFIX = ".." + string(filepath.Separator)
+
+func (fs *chroot) getRealPath(p string) (string, error) {
+	p = filepath.Join(fs.rootPath, p)
+	if fs.rootPath != "" && fs.rootPath != "/" {
+		rel, err := filepath.Rel(fs.rootPath, p)
```

Comment:
Sorry, but I didn't get the point, what's wrong here?

Reply:
You shouldn’t need to check if the `fs.rootPath != "/"` since `filepath.Rel(fs.rootPath, p)` would not generate an error, and the `strings.HasPrefix(rel, "../")` should never show up, and even if `filepath.Rel` didn’t error on `/../a/b` and instead just return `../a/b` we would want to return `os.ErrInvalid` the same.